### ISSUE TITLE
Collecting the symbolic constraints on input characters during concolic execution (question)

### ISSUE THREAD

**Issue by vrthra** [angr/angr#1198]
I have a program that takes a string (with some structure) as input. My aim is to collect information on what values each character in input is compared to as it gets parsed using concolic execution. What is the right way to go about it?

What I have tried so far:
#### Wrote a program that constructs an initial state with string arguments, executed it through stepping one at a time, and at the end, collect constraints.

```
import sys
import angr

class Program:
    def __init__(self, exe):
        self.exe = exe
        self.project = angr.Project(exe, load_options={'auto_load_libs': False})

    def set_input(self, arg):
        self.arg1 = arg

        self.initial_state = self.project.factory.entry_state(
                args=[self.exe, self.arg1],
                # does not seem to startup the unicorn engine
                add_options=angr.options.unicorn,
                )

    def run(self):
        state = self.initial_state
        while True:
            succ = state.step()
            if len(succ.successors) > 1:
                raise Exception('more successors %d' % len(succ.successors))
            if not succ.successors: return state
            state, = succ.successors

def main(exe, arg):
    prog = Program(exe)
    prog.set_input(arg)
    res = prog.run()
    print("constraints: %d" % len(res.solver.constraints))
    for i in res.solver.constraints:
        print(i)
    print('done')

if __name__ == '__main__':
    assert len(sys.argv) >= 3
    main(sys.argv[1], sys.argv[2])
```
When I execute this, I get zero constraints.

#### Using constrained symbolic arguments

I got the idea that I should make the arguments symbolic, and add constraints to each characters in the input, using state.add_constraints, and look for the constraints produced at the end.

```
import sys
import angr
import claripy

class Program:
    ARG_PREFIX = 'sym_arg'
    def __init__(self, exe):
        self.exe = exe
        self.project = angr.Project(exe, load_options={'auto_load_libs': False})

    def set_input(self, arg):
        # generate arg1 from individual characters.
        self.arg1 = self.make_symbolic_char_args(arg)

        self.initial_state = self.project.factory.entry_state(
                args=[self.exe, self.arg1],
                add_options=angr.options.unicorn,
                # does not seem to affect the number of constraints created
                # remove_options=angr.options.simplification
                )
        self.constrain_input_chars(self.initial_state, self.arg1a, arg)
        self.string_terminate(self.initial_state, self.arg1a, arg)

    def string_terminate(self, state, symarg, inarg):
        state.add_constraints(symarg[len(inarg)] == 0)

    def constrain_input_chars(self, state, symarg, sarg):
        constraints = []
        for i,a in enumerate(sarg):
            state.add_constraints(symarg[i] == a)

    def run(self):
        state = self.initial_state
        while True:
            succ = state.step()
            if len(succ.successors) > 1:
                raise Exception('more successors %d' % len(succ.successors))
            if not succ.successors: return state
            state, = succ.successors

    def make_symbolic_char_args(self, instr, symbolic=True):
        if not symbolic: return instr
        input_len = len(instr)
        largs = range(0, input_len+1)
        arg1k = ['%s_%d' % (Program.ARG_PREFIX, i) for i in largs]
        self.arg1h = {k:claripy.BVS(k, 8) for k in arg1k}
        self.arg1a = [self.arg1h[k] for k in arg1k]
        return reduce(lambda x,y: x.concat(y), self.arg1a)


def main(exe, arg):
    prog = Program(exe)
    prog.set_input(arg)
    res = prog.run()
    print("constraints: %d" % len(res.solver.constraints))
    for i in res.solver.constraints:
        print(i)
    print('done')

if __name__ == '__main__':
    assert len(sys.argv) >= 3
    main(sys.argv[1], sys.argv[2])
```

While this results in a large number of constraints at the end, I am at a loss how to interpret these constraints to what I want -- comparisons to input characters. The constraints are fairly complex, but there are lots duplicated portions in each item. Any advice on translating these constraints that look like below to actual character comparisons would be greatly appreciated.
```
...
<Bool (((if (sym_arg_0_0_8 == 0) then 0x0 else (if (sym_arg_1_1_8 == 0) then 0x1 else (if (sym_arg_2_2_8 == 0) then 0x2 else (if (sym_arg_3_3_8 == 0) then 0x3 else 0x4)))) + 0x7fffffffffeffd4) - 0x7fffffffffeffd4) == strlen_5_64>
<Bool 0x1 <= strlen_5_64>
<Bool ((0#56 .. 0#7 .. __invert__((if ((sym_arg_0_0_8 == 40) && (sym_arg_0_0_8[7:7] == 0)) then 1 else 0))) & 0xff) != 0x0>
<Bool (((if (sym_arg_0_0_8 == 0) then 0x0 else (if (sym_arg_1_1_8 == 0) then 0x1 else (if (sym_arg_2_2_8 == 0) then 0x2 else (if (sym_arg_3_3_8 == 0) then 0x3 else 0x4)))) + 0x7fffffffffeffd4) - 0x7fffffffffeffd4) == strlen_6_64>
<Bool 0x1 <= strlen_6_64>
<Bool ((0#56 .. 0#7 .. (if ((sym_arg_0_0_8 == 49) && (sym_arg_0_0_8[7:7] == 0)) then 1 else 0)) & 0xff) != 0x0>
<Bool (((if (sym_arg_0_0_8 == 0) then 0x0 else (if (sym_arg_1_1_8 == 0) then 0x1 else (if (sym_arg_2_2_8 == 0) then 0x2 else (if (sym_arg_3_3_8 == 0) then 0x3 else 0x4)))) + 0x7fffffffffeffd4) - 0x7fffffffffeffd4) == strlen_7_64>
<Bool 0x2 <= strlen_7_64>
<Bool ((0#56 .. 0#7 .. (if ((sym_arg_1_1_8 == 49) && (sym_arg_1_1_8[7:7] == 0)) then 1 else 0)) & 0xff) == 0x0>
<Bool (((if (sym_arg_0_0_8 == 0) then 0x0 else (if (sym_arg_1_1_8 == 0) then 0x1 else (if (sym_arg_2_2_8 == 0) then 0x2 else (if (sym_arg_3_3_8 == 0) then 0x3 else 0x4)))) + 0x7fffffffffeffd4) - 0x7fffffffffeffd4) == strlen_8_64>
<Bool 0x2 <= strlen_8_64>
...
```

##### Problem: too slow

The above approach also takes too long to execute even for inputs a few characters long. I tried executing the above on [simple recursive descent parser for expressions](https://github.com/vrthra/expr-tdr) with input of a few characters: (`1+4/1+4/(1+4)`). It took more than 30 minutes to parse!, while the running the parser directly comes out in less than a second. I would be OK with that time if there was no other way out, but this being concolic, I cant shake the feeling that there must be a better way.

#### Other options I have considered
* Writing a memory hook that monitors when the input string memory is read from.
  But this wont work for me because the characters may be copied before compared.

I have read the [angr docs](https://docs.angr.io/) multiple times now and can't figure out what to do. Any help would be greatly appreciated!.

**Comment by rhelmot** [comment]
So there are two pretty big issues that you're correct to be frustrated with:

1) angr is slow and this is a use case you can't use unicorn for
2) angr generates weird looking constraints

2 is caused partially because CPUs are horrible and do terrible things, partially because VEX is horrible and does terrible things, but quite significantly because when we ask z3 to simplify expressions, it produces expressions that are in fact easier and faster for it to reason about but horrifyingly complicated so far as a human would be concerned.

I don't think that 30 minutes is a reasonable amount of time for it to take for that example, though. Something else must be up. I'll look into it.

**Comment by rhelmot** [comment]
The 30 minutes is mostly because at every step, the only way we can decide which state should continue and which should not is by feeding it into z3, and the whole pipeline of shoving all the constraints down the FFI interface into z3 is very taxing. There's a lot of caching, but your example is very long and generates a lot of new symbols by calling strlen every time it breathes, so it exhausts the caches very quickly and individual steps start taking noticeable amounts of time at the constraint solving phase.

I did a quick rewrite to hopefully stay entirely out of z3, and it seems to be an order of magnitude faster. However, it's still a giant set of abstractions in python, which is slow.

With respect to the constraints being weird looking, there are exactly three kinds of constraints generated by your program:

- creation of strlen variables via large if-then-else trees so that these trees are not duplicated whenever a comparison against the strlen result is made
- comparisons against the strlen variables (every time you call next_token it adds another of these)
- constraints on the individual characters. These look weird because the compiler produces a weird series of operations in order to do the comparisons - swapping the value between rax, eax, and al produces sign/zero extends and extracts, and there is in fact a xor operation in the assembly.

```python
import sys
import angr
import claripy

class Program:
    ARG_PREFIX = 'sym_arg'
    def __init__(self, exe):
        self.exe = exe
        self.project = angr.Project(exe, load_options={'auto_load_libs': False})

    def set_input(self, arg):
        # generate arg1 from individual characters.
        self.arg1 = self.make_symbolic_char_args(arg)

        # use LAZY_SOLVES to stay out of z3
        self.initial_state = self.project.factory.entry_state(
                args=[self.exe, self.arg1],
                add_options={angr.options.LAZY_SOLVES},
                remove_options=angr.options.simplification,
                )
        self.constrain_input_chars(self.initial_state, self.arg1a, arg)
        self.string_terminate(self.initial_state, self.arg1a, arg)

    def string_terminate(self, state, symarg, inarg):
        self.assignments[symarg[len(inarg)].cache_key] = claripy.BVV('\0')

    def constrain_input_chars(self, state, symarg, sarg):
        self.assignments = {}
        for i,a in enumerate(sarg):
            self.assignments[symarg[i].cache_key] = claripy.BVV(a)

    def run(self):
        state = self.initial_state
        passed = set()
        while True:
            succ = state.step()

            # do our own smart constraint solving since z3 is a blunt instrument
            successors = []
            for succ_state in succ.successors:
                for c in succ_state.solver.constraints:
                    # if we've already validated this don't bother
                    if c.cache_key in passed:
                        continue

                    # this might be the creation of a new symbol we need to register the value of
                    if c.op == '__eq__':
                        new_sym = None
                        new_val = None
                        new_val_conc = None
                        if c.args[0].op == 'BVS' and c.args[0].cache_key not in self.assignments:
                            new_sym, new_val = c.args
                        elif c.args[1].op == 'BVS' and c.args[1].cache_key not in self.assignments:
                            new_val, new_sym = c.args
                        if new_sym is not None:
                            if len(succ.successors) != 1:
                                raise Exception("I was not prepared for this")
                            new_val_conc = new_val.replace_dict(self.assignments)
                            if new_val_conc.op != 'BVV':
                                raise Exception("Couldn't come up with value for new symbol")
                            self.assignments[new_sym.cache_key] = new_val_conc
                            passed.add(c.cache_key)
                            continue

                    # try evaluating the constraint with our current assignments
                    better_constraint = c.replace_dict(self.assignments)
                    if better_constraint.is_false():
                        break
                    if not better_constraint.is_true():
                        raise Exception("Could not simplify constraint to true/false")
                    passed.add(c.cache_key)
                else:
                    successors.append(succ_state)

            if len(successors) > 1:
                raise Exception('more successors %d' % len(successors))
            if not successors: return state
            state, = successors

    def make_symbolic_char_args(self, instr, symbolic=True):
        if not symbolic: return instr
        input_len = len(instr)
        largs = range(0, input_len+1)
        arg1k = ['%s_%d' % (Program.ARG_PREFIX, i) for i in largs]
        self.arg1h = {k:claripy.BVS(k, 8) for k in arg1k}
        self.arg1a = [self.arg1h[k] for k in arg1k]
        return reduce(lambda x,y: x.concat(y), self.arg1a)

def main(exe, arg):
    prog = Program(exe)
    prog.set_input(arg)
    res = prog.run()
    print("constraints: %d" % len(res.solver.constraints))
    for i in res.solver.constraints:
        print(i)
    print('done')

if __name__ == '__main__':
    assert len(sys.argv) >= 3
    main(sys.argv[1], sys.argv[2])
```

In summary: Welcome to binary analysis, it sucks down here.
